### PR TITLE
feat: add purity annotation

### PR DESCRIPTION
**Pull request by bollu**
Add a custom field into each `Op` that declares if its signature is pure or not. If it is pure, it returns `a`, and if impure, it returns `IO a`. The default signature object is pure.

### CODE REVIEW

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:1235]: leftover from a merge?

File: SSA/Projects/InstCombine/Tactic.lean
Comment (goens) [inline SSA/Projects/InstCombine/Tactic.lean:23]: in the paper we call it `LLVM` to add to the confusion :grinning:

**Comment by alexkeizer** [comment]
I merged main, and bumped the heartbeats one more, now it seems to be working locally

**Comment by bollu** [comment]
Nice! I flipped the state to ready to review. Since alex and I coauthored this, I think we need sign off from @tobiasgrosser @goens on this.

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:140]:
```suggestion
/-- An intrinsically typed expression whose effect is *at most* EffectKind -/
```

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:626]: I don't understand this TODO. What's preventing this from being `@[simp]`?

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:666]: I'd like to know as well :)

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:726]: stray code? or something that needs to be finished later? (can we either uncomment or delete it?)

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:743]: What's the issue with this lemma? Why is it commented out?

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:760]: Why is this a TODO and not a change? What's the issue with doing it?

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:770]: Should we do this one too before merging? (seems simple enough)

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:776]: maybe we should do this before merging?

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:953]: Can we uncomment or delete these?

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:1053]: IMO we should probably split this file up anyway (but that's maybe a question for later)

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:1573]: Why are we keeping this around? (It might be worth having the counterexample be an `example` and made explicit, I guess?)

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:1582]:
```suggestion
```

File: SSA/Core/Framework.lean
Comment (goens) [inline SSA/Core/Framework.lean:2052]:
```suggestion
/- TODO: we might not need `denoteIntoSubtype`, if we can prove that `V ∈ supp (lets.denote _)`
```

**Review by goens — APPROVED** [review]
Thanks @alexkeizer and @bollu for the heroic effort here! It's nice that you've also documented things well in this PR. Overall, I'm happy with it. I've left a few comments here that are small to address. 

There's a bunch of TODO's introduced by this PR, and it's certainly not worth doing them all right now. It would be nice to have maybe an issue tracking the cleanup of this and all the TODO's, so we don't forget to do it afterwards?

File: SSA/Core/EffectKind.lean
Comment (bollu) [reply SSA/Core/EffectKind.lean:171]: @goens , @alexkeizer told me that there are situations where proof are not "really irrelevant", so we take the theorem as a parameter. Maybe we can make this a default argument instead?

File: SSA/Core/Examples.lean
Comment (bollu) [reply SSA/Core/Examples.lean:32]: Agreed, Tracked by issue https://github.com/opencompl/ssa/issues/250

File: SSA/Core/Examples.lean
Comment (bollu) [reply SSA/Core/Examples.lean:404]: Tracked at https://github.com/opencompl/ssa/issues/251

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:70]: Tracked at https://github.com/opencompl/ssa/issues/252

File: SSA/Core/EffectKind.lean
Comment (goens) [reply SSA/Core/EffectKind.lean:171]: Oh is this about `simp` and how the proof looks like syntactically? It would be nice to understand that better, but I don't want this to block this merge. I made an issue (#253) to keep track of this one.

**Comment by tobiasgrosser** [comment]
I think we should address https://github.com/opencompl/ssa/issues/250 in this very PR, if possible, as it will minimize the diff of this PR quite a bit.

**Comment by alexkeizer** [comment]
I'm not sure we should block this PR on #250, since the solution is not uncontroversially clear (unless @bollu agrees with my counter-proposal)

File: SSA/Core/EffectKind.lean
Comment (bollu) [reply SSA/Core/EffectKind.lean:32]: @goens principle of least privilege :) One can call `toMonad` even when `m` is not necessarily a monad. So, I'd rather not add a `[Monad m]` constraint on `toMonad`.

File: SSA/Core/EffectKind.lean
Comment (bollu) [reply SSA/Core/EffectKind.lean]: Resolved, removed the deprecated `liftEffect_pure_pure`.

File: SSA/Core/EffectKind.lean
Comment (bollu) [reply SSA/Core/EffectKind.lean:166]: Resolving conversation.

File: SSA/Core/ErasedContext.lean
Comment (bollu) [reply SSA/Core/ErasedContext.lean:466]: @goens I am not sure what this comment means.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean]: As the author of the comment, neither do I, so I nuked it.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:681]: As the author of the comment, neither do I, so I deleted the comment :D

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean]: IIRC, this was a proof strategy that we did not pursue. Removed for the sake of cleanliness.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean]: Proof strategy that was not pursused. Removed for cleanliness.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:755]: done.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:771]: done.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:765]: done.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:1041]: Yes, let's do this post-merge, please!

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:1417]: @goens This is a good refactor, but to be consistent, we would need to change a *lot* of names of lemmas, theorems, and other definitions that use the `vars` terminology right now. I'd suggest doing this after.

File: SSA/Core/EffectKind.lean
Comment (bollu) [reply SSA/Core/EffectKind.lean:166]: Resolving, since we are tracking this as an issue now.

File: SSA/Core/Framework.lean
Comment (bollu) [reply SSA/Core/Framework.lean:1323]: done.

File: SSA/Core/ErasedContext.lean
Comment (bollu) [reply SSA/Core/ErasedContext.lean:466]: Resolved over at Zulip

**Pull request merged**
